src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+    {
+        auto chPos = chExtBeg;
+        auto it = _charOffsets.begin() + colExtBeg;
+
+        it = iota_n_mut(it, leadingSpaces, chPos);
+
+        *it++ = chPos;
+        it = fill_small(it, _charOffsets.begin() + colEnd, gsl::narrow_cast<uint16_t>(chPos | CharOffsetsTrailer));
+        chPos = gsl::narrow_cast<uint16_t>(chPos + chars.size());
+
+        it = iota_n_mut(it, trailingSpaces, chPos);
+    }
+}
+
+const til::small_rle<TextAttribute, uint16_t, 1>& ROW::Attributes() const noexcept
+{
+    return _attr;
+}
+
+TextAttribute ROW::GetAttrByColumn(const til::CoordType column) const
+{
+    return _attr.at(_clampedUint16(column));
+}
+
+std::vector<uint16_t> ROW::GetHyperlinks() const
+{
+    std::vector<uint16_t> ids;
+    for (const auto& run : _attr.runs())
+    {
+        if (run.value.IsHyperlink())
+        {
+            ids.emplace_back(run.value.GetHyperlinkId());
+        }
+    }
+    return ids;
+}
+
+uint16_t ROW::size() const noexcept
+{
+    return _columnCount;
+}
+
+til::CoordType ROW::MeasureLeft() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            break;
+        }
+    }
+
+    return gsl::narrow_cast<til::CoordType>(it - beg);
+}
+
+til::CoordType ROW::MeasureRight() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = end;
+
+    for (; it != beg; --it)
+    {
+        // it[-1] is safe as `it` is always greater than `beg` (loop invariant).
+        if (til::at(it, -1) != L' ')
+        {
+            break;
+        }
+    }
+
+    // We're supposed to return the measurement in cells and not characters
+    // and therefore simply calculating `it - beg` would be wrong.
+    //
+    // An example: The row is 10 cells wide and `it` points to the second character.
+    // `it - beg` would return 1, but it's possible it's actually 1 wide glyph and 8 whitespace.
+    return gsl::narrow_cast<til::CoordType>(_columnCount - (end - it));
+}
+
+bool ROW::ContainsText() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+std::wstring_view ROW::GlyphAt(til::CoordType column) const noexcept
+{
+    auto col = _clampedColumn(column);
+
+    // Safety: col is [0, _columnCount).
+    const auto beg = _uncheckedCharOffset(col);
+    // Safety: col cannot be incremented past _columnCount, because the last
+    // _charOffset at index _columnCount will never get the CharOffsetsTrailer flag.
+    while (_uncheckedIsTrailer(++col))
+    {
+    }
+    // Safety: col is now [1, _columnCount].
+    const auto end = _uncheckedCharOffset(col);

```

Comment:
Don't we want `col=0` to be valid though?

Reply:
`col` can only be `1` at that point because the `while` loop contains `++col`.
This is important to know, because it shifts the numeric range of `col` from `[0, _columnCount)` to `(0, _columnCount]` and `_columnCount` is the absolute maximum you can use to index into the `_charOffsets` buffer. This proves that this isn't resulting in a out of bounds read.

Reply:
Oh I see it now, my bad. I thought `beg` was bound to `(0,_columnCount]` which made me worried that `GlyphAt(0)` would look at `1` instead. We're good!